### PR TITLE
Add workers pool to pubsub

### PR DESCRIPTION
**Pull request by aleklisi**
This PR addresses introduces a wpool to handle iqs in pubsub.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[6167.1](https://travis-ci.org/esl/MongooseIM/jobs/488451754) / Erlang 19.3 / small_tests / 44d408f52ffac8339c67633ea00282d2a3b23f3f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2212/6167/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2212/6167/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-02-04_11.37.14/lib.mongooseim.logs/run.2019-02-04_11.37.15/suite.log.html)

File: big_tests/tests/pubsub_SUITE.erl [hunk 1]
```diff
@@ -644,11 +644,15 @@ retract_when_user_goes_offline_test(Config) ->
...
               pubsub_tools:get_all_items(Alice, Node,
                                          [{expected_result, [<<"item2">>, <<"item1">>]}]),
-
               escalus_client:stop(Config, Bob),
-              pubsub_tools:get_all_items(Alice, Node, [{expected_result, [<<"item1">>]}]),
-
-              pubsub_tools:delete_node(Alice, Node, [])
+              mongoose_helper:wait_until(
+                fun() ->
+                      [#{id => ok}] == pubsub_tools:get_all_items(Alice, Node, [{expected_result, [<<"item1">>]}])
```
Comment (zofpolkowska) [inline big_tests/tests/pubsub_SUITE.erl]:
```suggestion
                      fun() ->
                         pubsub_tools:get_all_items(Alice, Node, [{expected_result, [<<"item1">>]}])
                      end,
                      [#{id => ok}],
                      ...
```

Comment (zofpolkowska) [reply big_tests/tests/pubsub_SUITE.erl]: Small improvement :)

File: src/pubsub/mod_pubsub.erl [hunk 2]
```diff
@@ -99,6 +93,11 @@
 
 -export([send_loop/1]).
 
+%% for pubsub_router
+-export([
+    do_route/7
```
Comment (zofpolkowska) [inline src/pubsub/mod_pubsub.erl]: Maybe it could be moved to the `pubsub_router` module ?

File: src/pubsub/mod_pubsub.erl [hunk 3]
```diff
@@ -228,24 +195,44 @@ start(Host, Opts) ->
     ChildSpec = {Proc, {?MODULE, start_link, [Host, Opts]},
                  transient, 1000, worker, [?MODULE]},
     ensure_metrics(Host),
-    ejabberd_sup:start_child(ChildSpec).
+    R = ejabberd_sup:start_child(ChildSpec),
+    %% Wpool would call mongoose_pubsub_router:init with InitOpts
```
Comment (zofpolkowska) [inline src/pubsub/mod_pubsub.erl]: Not necessary comment

File: src/pubsub/pubsub_index.erl [hunk 4]
```diff
@@ -41,7 +41,18 @@ init() ->
                          {attributes, record_info(fields, pubsub_index)}]).
 
 new(Index) ->
-    case mnesia:read({pubsub_index, Index}) of
+    Parent = self(),
```
Comment (zofpolkowska) [inline src/pubsub/pubsub_index.erl]: This is not needed if we don't use `worker_pool` for mnesia

File: big_tests/tests/pubsub_SUITE.erl (hunk 1 above)
Comment (aleklisi) [reply big_tests/tests/pubsub_SUITE.erl]: fixed

File: src/pubsub/mod_pubsub.erl (hunk 3 above)
Comment (aleklisi) [reply src/pubsub/mod_pubsub.erl]: fixed

File: src/pubsub/pubsub_index.erl (hunk 4 above)
Comment (aleklisi) [reply src/pubsub/pubsub_index.erl]: fixed

**Comment by mongoose-im** [comment]
[6173.1](https://travis-ci.org/esl/MongooseIM/jobs/488509932) / Erlang 19.3 / small_tests / ec5e3973a59f402bf5122240b5740b31844fdae9
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2212/6173/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2212/6173/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-02-04_14.17.39/lib.mongooseim.logs/run.2019-02-04_14.17.40/suite.log.html)

**Comment by aleklisi** [comment]
This PR turns out to work worse than #2206

**Pull request closed without merging**